File: Apps/Chat/android/chatkit/src/main/kotlin/com/dweekly/cyrinx/chat/SplitMix64.kt
package com.dweekly.cyrinx.chat

/**
 * SplitMix64, the seeded PRNG pinned for the simulated chat client
 * (../../../CONTRACT.md section 2, point 4). Public-domain reference algorithm:
 * Vigna, `splitmix64.c` <http://prng.di.unimi.it/splitmix64.c>; also the generator
 * from Steele, Lea & Flood, "Fast Splittable Pseudorandom Number Generators,"
 * OOPSLA 2014.
 *
 * `state` is a Kotlin `Long` used purely as a 64-bit bit-pattern container: all
 * arithmetic below (`+`, `*`) is taken modulo 2^64, which is exactly what JVM `Long`
 * addition/multiplication already does (two's-complement wraparound is bit-for-bit
 * identical whether the pattern is interpreted as signed or unsigned), and all
 * right-shifts use `ushr` (logical/unsigned shift), matching the reference
 * algorithm's unsigned `>>`. No behavior here depends on `Long` being signed.
 *
 * Constructed directly from the scenario's `seed` with no additional hashing of the
 * seed itself, per ../../../CONTRACT.md section 2.
 */
class SplitMix64(seed: Long) {
    // Per-call state increment ("golden gamma"), pinned in CONTRACT.md section 2.
    // Hex literal written via the ULong->Long bit-reinterpreting conversion because
    // 0x9E3779B97F4A7C15 has its top bit set and does not fit as a positive Kotlin
    // Long literal.
    private var state: Long = seed

    /** One u64 draw, per the reference `next()` step pinned in
     * ../../../CONTRACT.md section 2. */
    fun next(): Long {
        state += GOLDEN_GAMMA
        var z = state
        z = (z xor (z ushr 30)) * MIX_MUL_1
        z = (z xor (z ushr 27)) * MIX_MUL_2
        z = z xor (z ushr 31)
        return z
    }

    companion object {
        // Written as ULong hex literals then bit-reinterpreted to Long via toLong()
        // (rather than hand-converted to a signed decimal constant) so the source
        // shows the exact hex value from CONTRACT.md section 2 with no manual
        // two's-complement arithmetic to get wrong.

        /** Per-call state increment. ../../../CONTRACT.md section 2. */
        val GOLDEN_GAMMA: Long = 0x9E3779B97F4A7C15uL.toLong()

        /** First avalanche multiplier. ../../../CONTRACT.md section 2. */
        val MIX_MUL_1: Long = 0xBF58476D1CE4E5B9uL.toLong()

        /** Second avalanche multiplier. ../../../CONTRACT.md section 2. */
        val MIX_MUL_2: Long = 0x94D049BB133111EBuL.toLong()
    }
}

/** Encodes a `Long` as its 8-byte big-endian representation, matching "Draw N -> 8
 * bytes (big-endian) of the u64 result" in ../../../CONTRACT.md section 2. */
fun Long.toBigEndianBytes(): ByteArray {
    val out = ByteArray(8)
    for (i in 0 until 8) {
        out[i] = ((this ushr (8 * (7 - i))) and 0xFF).toByte()
    }
    return out
}

File: Apps/Chat/CyrinxChatKit/Sources/CyrinxChatKit/SplitMix64.swift
/// Deterministic seeded PRNG for `SimulatedChatTransportClient`.
/// Apps/Chat/CONTRACT.md §2, point 4.
///
/// Public-domain reference algorithm: Vigna, `splitmix64.c`,
/// <http://prng.di.unimi.it/splitmix64.c>; also the generator described in
/// Steele, Lea & Flood, "Fast Splittable Pseudorandom Number Generators,"
/// OOPSLA 2014. Constants and the `next()` step below are transcribed
/// verbatim from CONTRACT.md §2, which itself transcribes them from the
/// cited reference -- they are not free parameters and must not be
/// "improved" independently of a spec change.
public struct SplitMix64: Sendable {
    /// Per-call state increment (CONTRACT.md §2).
    static let goldenGamma: UInt64 = 0x9E37_79B9_7F4A_7C15
    /// First avalanche multiplier (CONTRACT.md §2).
    static let mixMul1: UInt64 = 0xBF58_476D_1CE4_E5B9
    /// Second avalanche multiplier (CONTRACT.md §2).
    static let mixMul2: UInt64 = 0x94D0_49BB_1331_11EB

    private var state: UInt64

    /// `state` is initialized directly from `seed`, no additional hashing
    /// of the seed itself (CONTRACT.md §2).
    public init(seed: UInt64) {
        self.state = seed
    }

    /// One u64 draw. Reference step, identical on both platforms
    /// (CONTRACT.md §2):
    /// ```
    /// state = state + GOLDEN_GAMMA            // (mod 2^64)
    /// z = state
    /// z = (z ^ (z >> 30)) * MIX_MUL_1         // (mod 2^64)
    /// z = (z ^ (z >> 27)) * MIX_MUL_2         // (mod 2^64)
    /// z = z ^ (z >> 31)
    /// return z
    /// ```
    public mutating func next() -> UInt64 {
        state = state &+ Self.goldenGamma
        var z = state
        z = (z ^ (z >> 30)) &* Self.mixMul1
        z = (z ^ (z >> 27)) &* Self.mixMul2
        z = z ^ (z >> 31)
        return z
    }
}

/// Big-endian byte decomposition of a `UInt64`, most-significant byte
/// first. Used by the simulator's PRNG draw-order contract (CONTRACT.md §2:
/// "Draw 1 -> 8 bytes (big-endian) of the u64 result, of which the first 4
/// bytes become client A's simulated local peer ID").
func splitMix64BigEndianBytes(_ value: UInt64) -> [UInt8] {
    (0..<8).map { index in
        UInt8((value >> (56 - 8 * UInt64(index))) & 0xFF)
    }
}
